test/functional/change_stream.test.js

```diff
@@ -2017,11 +2017,7 @@ describe('Change Streams', function() {
           return Promise.resolve()
             .then(() => changeStream.next())
             .then(() => changeStream.next())
-            .then(() => {
-              const nextP = changeStream.next();
-
-              return changeStream.close().then(() => nextP);
-            });
+            .then(() => Promise.all([changeStream.close(), changeStream.next()]));
```

Comment:
This is a behavioral change, it used to be that `next` was called (executing soon), and then the `close` call was made, after which next was awaited. Now you're essentially racing both `close` and `next`, and `close` comes first. 

Was there something you're working around here?

Reply:
I thought I wrote a comment about that but I don't see it anymore, odd. 

> it used to be that next was called (executing soon), and then the close call was made, after which next was awaited

The test fails under those conditions. The final next returns the change when it is awaited, and the test expects an error. I can only get it to return an error if I call `close` first. I'm still not exactly sure what the desired behavior is that this test is trying to confirm happens.

Reply:
Here's the pull request where it was introduced, for more context: https://github.com/mongodb/node-mongodb-native/pull/1902

I think the idea is that we want to test the behavior when a new change comes in while the user is actively closing the change stream. It might just be an edge case that was possible before some fix that you introduced in this PR? It also seems like a test that could only really be handled through mocking or clever use of fail points, do you think we should remove the test?

Reply:
Yeah this is all making more sense now. I think the difference in behavior is derived from [this change](https://github.com/mongodb/node-mongodb-native/pull/2294/files#diff-30861ceb96e5170099331504d187951eR489), at which point they started flaking out until I skipped them in #2322. 

If I'm understanding correctly, this test was checking to make sure that `processNewChange` would throw an error if the underlying cursor closed. However, that shouldn't really be the case - the underlying cursor closes during a resumable error, but the stream shouldn't error.

So yeah, if that's correct I'm all for removing the test.

Reply:
I _think_ I agree, but now I'm wondering more about the `isClosed` dilemma also addressed in this PR. `processNewChange` has:
```
  if (cursor == null) {
    changeStream.closed = true;
    if (eventEmitter) {
      return;
    }

    callback(new MongoError('ChangeStream is closed'));
    const error = new MongoError('ChangeStream is closed');
    return;
  }
```

But what happens if `changeStream.closed === true` (i.e. the change stream was explicitly closed)? Shouldn't that trigger this path as well? Is that the behavior this test is testing?

Reply:
`next` already checks if `changeStream.closed === true` and returns an error before calling `processNewChange`. I confirmed by adding this to `processNewChange` and the test suite ran with no errors.
```
  if (changeStream.closed) {
    throw new Error('Should never get here');
  }
```

Reply:
That said, I'm thinking I should flip the or condition in `isClosed`.

If it was `return this.closed || (this.cursor && this.cursor.isClosed());` instead of the other way around, `closed === true` would take precedence over the cursor state.

Reply:
`next` checks this before calling `next` on the cursor, which in turn calls `processNewChange`. What if the change stream is not closed when you call `next`, but is closed sometime before the result of `cursor.next` returns? You might be able to test this with a mock server (or a sinon mock if we had a more complex fixture in place):
```
server.setResponseHandler(res => {
   const doc = res.doc;
   if (doc.ismaster) {
     res.reply({ .. ismaster .. });
   } else if (doc.aggregate) {
     res.reply({ ok: 1, .. other aggregate stuff .. });
   } else if (doc.getMore) {
     setTimeout(() => res.reply({ ok: 1, .. other getmore stuff .. }, 200); // always respond with latency to `getMore` calls
   }
});
```

now what does this do:
```
const changeStream = coll.watch();
changeStream.next(err => { // first call will send `aggregate`
  expect(err).to.not.exist;

  changeStream.next(err => {  // second call will send `getMore`
    expect(err).to.exist; // I think we _do_ expect to get an error here, change stream was closed explicitly. 
    done();
  });

  changeStream.close(); // immediately close the change stream
});
```

It may be that the underlying cursor logic already protects us from this, but I think that this is the behavior the test is meaning to cover. I just want to make sure we're not losing that coverage here.

Reply:
That makes sense, I'll check into it.

Reply:
I'm thinking it would make sense to create a separate ticket to fix those two `mid-close` tests. They were already skipped before this PR, so keeping them skipped until a followup PR doesn't hurt IMO. Thoughts, @mbroadst?